monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/IntersperseObservable.scala

```diff
@@ -0,0 +1,56 @@
+package monix.reactive.internal.operators
+
+import monix.execution.Ack.Continue
+import monix.execution.atomic.AtomicBoolean
+import monix.execution.{Ack, Cancelable}
+import monix.reactive.Observable
+import monix.reactive.observers.Subscriber
+
+import scala.concurrent.Future
+
+/**
+  * Created by omainegra on 5/29/17.
+  */
+private[reactive] final class IntersperseObservable[+A](source: Observable[A],
+                                                        separator: A) extends Observable[A]{ self =>
+
+  override def unsafeSubscribeFn(out: Subscriber[A]): Cancelable = {
+    val firstTime = AtomicBoolean(true)
+    val isDone = AtomicBoolean(false)
+    var downstreamAck = Continue : Future[Ack]
+
+    val upstream = source.unsafeSubscribeFn(new Subscriber[A] {
+      implicit val scheduler = out.scheduler
+
+      override def onNext(elem: A): Future[Ack] = {
+        downstreamAck = if (firstTime.getAndSet(false)) {
+          out.onNext(elem)
+        }
+        else {
+          out.onNext(separator).flatMap {
+            case Continue => out.onNext(elem)
+            case ack => ack
+          }
+        }
+        downstreamAck
+      }
+
+      def onError(ex: Throwable) = {
+        if (!isDone.getAndSet(true)){
```

Comment:
Just as with `firstTime`, you're only modifying `isDone` in `onComplete` and `onError`. The calls to `onNext`, `onComplete` and `onError` are ordered (synchronized) already so this can be a `private[this] var`.

The only concurrency problem you can have is that the producer doesn't need to back-pressure for that final `Future[Ack]` returned by the last `onNext`, so depending what you do in those `flatMaps`, you can end up with a concurrency problem, but you don't need anything special here.